### DISCUSSION TITLE
Protocol Buffer based messages

### DISCUSSION THREAD

**Discussion by pascalwhoop** [united-manufacturing-hub/united-manufacturing-hub#459] [Q&A] [answered]
Hey,
have you considered defining all non-raw messages as protocol buffer files? It gives a lot of type safety that can be leveraged by downstream consumers. Just curious

**Comment by Gornoka** [comment]
> Hey,
> have you considered defining all non-raw messages as protocol buffer files? It gives a lot of type safety that can be leveraged by downstream consumers. Just curious
> 
> 

We discussed going a step further, creating an async API specification for the entire mqtt environment (https://www.asyncapi.com/), but have not reached a conclusion yet.  However protocol buffer files may be another option to receive a similar level of stability, at least for the message contents.

**Comment by JeremyTheocharis** [comment, marked as answer]
Actually, we thought about using protobuf as a second option!

Another advantage: it saves data, especially on mobile connections.

Disadvantage: it introduces complexity in implementing or maintaining new production machines. This task is usually done by production engineers without Programming experience. Protobuf might be too complicated for them.

Conclusion: introduce protobuf as a second option to improve overall stability when the system is more mature, e.g., more pre-defined Microservices that reduce the need for Node-RED.
